File: 02_data_type/main.py
# Number, String, List, Dictionary, Tuple, Set, NoneType

# Number(숫자형: int, float, bool)
x = 100
y = 400.1
z = True

print(f"type of x = {type(x)}")  # type of x=<class 'int'>
print(f"type of y = {y.__class__}")  # type of y=<class 'float'>
print(f"type of y's class = {y.__class__.__class__}")  # type of y's class=<class 'type'>
print(f"is z an int? {isinstance(z, int)}")  # is z an int? True
print(f"hexadecimal address of x = {hex(id(x))}")  # hexadecimal address of x=0x...
print(f"x is y ? {x is y}")  # x is y? False

# String(문자열: str)
a = "안녕하세요"
print(f"first character of a = {a[0]}")  # first character of a=안
print(f"length of a = {len(a)}")  # length of a=5

# List(리스트: list, 대괄호 안에 객체들을 순서대로 모아놓은 객체)
X = [
    {
        (1, 2, 3): ["강물", "야구", [100, 300]],
        "취미": ["스쿠버", "수영"]
    }
]
print(f"element in X = {X[0][(1, 2, 3)][2][1]}")  # element in X=300

# Dictionary(사전: dict, 중괄호 안에 객체들을 랜덤 순서대로 모아놓은 객체)
Y = {
    "age": 36,
    "name": "류종훈",
    "hobby": ["스쿠버", "수영", "경연프로그램 보기"],
    "last name": "Jasper",
    0: "0",
    1: 1
}
print(f"element in Y = {Y[0]}")  # element in Y=0
print(f"age in Y = {Y['age']}")  # age in Y=36
print(f"length of Y={len(Y)}")  # length of Y=5

# Tuple(튜플: tuple, 괄호 안에 객체들을 왼쪽에서 오른쪽 순서대로 모아놓은 객체)
Z = ("안녕하세요",)
b = ([1, 2, 3], (4, 5), {"이름": "류종훈", "나이": 36}, "한국")
print(f"is Z a tuple ? {isinstance(Z, tuple)}")  # is Z a tuple? True
print(f"class of Z = {Z.__class__}")  # class of Z=<class 'tuple'>
print(f"name in b = {b[2]['이름']}")  # name in b=류종훈
print(f"length of b = {len(b)}")  # length of b=4

# Set(세트: set, 중괄호 안에 객체들을 랜덤순서대로 모아 놓은 객체)
c = {"안녕하세요"}
print(f"type of c = {type(c)}")  # type of c=<class 'set'>

# NoneType(None, 존재하지 않는 또는 값이 없다를 정의하는 유형, 에러방지용 초기화에 사용)
d = None
print(f"type of d = {type(d)}")  # type of d=<class 'NoneType'>

File: 02_data_type/main.cpp
#include <iostream>
#include <typeinfo>

using namespace std;

int main() {
    // Number(숫자형: int, float, bool)
    int x = 100;
    float y = 400.1;
    bool z = true;

    cout << "type of x = " << typeid(x).name() << endl;  // type of x=i
    cout << "type of y = " << typeid(y).name() << endl;  // type of y=f
    cout << "type of y's class = " << typeid(typeid(y)).name() << endl;  // type of y's class=St9type_info
    cout << "is z an int? " << (z ? "true" : "false") << endl;  // is z an int? true
    cout << "hexadecimal address of x = " << &x << endl;  // hexadecimal address of x=0x...
    cout << "x is y ? " << (x == y ? "true" : "false") << endl;  // x is y? False

    // Char(문자형: char, string과 char 배열 사용 가능)
    char ch = 'A';
    string str = "Hello";
    char charArray[] = "World";
    cout << "char value: " << ch << endl;  // char value: A
    cout << "string value: " << str << endl;  // string value: Hello
    cout << "char array value: " << charArray << endl;  // char array value: World
    
    // String(문자열: str)
    string a = "Hello";
    cout << "first character of a = " << a[0] << endl;  // first character of a=안
    cout << "length of a = " << a.length() << endl;  // length of a=5

    // List(리스트: list, 대괄호 안에 객체들을 순서대로 모아놓은 객체)
    // C++에는 리스트와 사전과 같은 직접적인 타입이 없으므로 이 예제는 생략합니다.

    // Dictionary(사전: dict, 중괄호 안에 객체들을 랜덤 순서대로 모아놓은 객체)
    // C++에는 사전과 같은 직접적인 타입이 없으므로 이 예제는 생략합니다.

    // Tuple(튜플: tuple, 괄호 안에 객체들을 왼쪽에서 오른쪽 순서대로 모아놓은 객체)
    // C++에는 튜플과 같은 직접적인 타입이 없으므로 이 예제는 생략합니다.

    // Set(세트: set, 중괄호 안에 객체들을 랜덤순서대로 모아 놓은 객체)
    // C++에는 세트와 같은 직접적인 타입이 없으므로 이 예제는 생략합니다.



    // List의 경우 int my_array[] = 방식으로 사용할 수 있다.
    int my_array[] = {1, 2, 3, 4, 5};
    cout << "element in my_array = " << my_array[2] << endl;  // element in my_array = 3

    // Sizeof 예제
    cout << "size of x (in bytes) = " << sizeof(x) << endl;  // size of x (in bytes) = 4
    cout << "size of y (in bytes) = " << sizeof(y) << endl;  // size of y (in bytes) = 4
    cout << "size of ch (in bytes) = " << sizeof(ch) << endl;  // size of ch (in bytes) = 1

    // NoneType(None, 존재하지 않는 또는 값이 없다를 정의하는 유형, 에러방지용 초기화에 사용)
    nullptr_t d = nullptr;
    cout << "type of d = " << typeid(d).name() << endl;  // type of d=std::nullptr_t

    return 0;
}
